File: comparative_tests/go_tests/image_manipulation_functions/grayscale_filter.go
package image_manipulation_functions

import (
	"fmt"
	bft "go_tests/basic_functions"
	"image"
	"image/color"
	"time"
)

func GrayscaleFilter(img image.Image, intensity int, outputsPath string) {
	start := time.Now()

	// Ensure the output directory exists
	if err := EnsureOutputDir(outputsPath); err != nil {
		fmt.Println(err)
		return
	}

	// Get the image dimensions
	bounds := img.Bounds()
	width, height := bft.GetImageDimensions(img)

	grayedImage := image.NewRGBA(bounds)

	// Apply the grayscale filter
	for y := 0; y < height; y++ {
		for x := 0; x < width; x++ {
			originalColor := color.RGBAModel.Convert(img.At(x, y)).(color.RGBA)

			// Calculate the luminosity of the pixel
			grayValue := uint8(0.3*float64(originalColor.R) + 0.59*float64(originalColor.G) + 0.11*float64(originalColor.B))

			// Apply the grayscale filter with the intensity (0 = no change, 100 = full grayscale)
			blendFactor := float64(intensity) / 100
			r := uint8((1-blendFactor)*float64(originalColor.R) + blendFactor*float64(grayValue))
			g := uint8((1-blendFactor)*float64(originalColor.G) + blendFactor*float64(grayValue))
			b := uint8((1-blendFactor)*float64(originalColor.B) + blendFactor*float64(grayValue))

			// Set the new pixel color
			grayedImage.Set(x, y, color.RGBA{
				R: r,
				G: g,
				B: b,
				A: originalColor.A,
			})
		}
	}

	// Create the output file
	outputFile, err := CreateOutputFile(outputsPath)
	if err != nil {
		fmt.Println(err)
		return
	}

	// Encode the grayed image as JPEG and save it to the output file
	if err := SaveImageAsJPEG(outputFile, grayedImage); err != nil {
		fmt.Println(err)
		return
	}

	fmt.Println("Grayed image saved to : ", outputsPath)
	fmt.Println("Execution time : ", time.Since(start))
}

File: comparative_tests/python_tests/image_manipulation_functions/grayscale_filter.py
from PIL import Image
import numpy as np
from . import utils
import time


def grayscale_filter(img, intensity, outputs_path):
    start_time = time.time()

    # Ensure the output directory exists
    utils.ensure_output_dir(outputs_path)

    # Conversion of the image to a numpy array
    img_array = np.array(img.convert("RGBA"))

    # Apply the grayscale filter
    grayscale_array = img_array.copy()
    blend_factor = intensity / 100.0

    for y in range(img_array.shape[0]):
        for x in range(img_array.shape[1]):
            original_color = img_array[y, x]
            gray_value = int(0.3 * original_color[0] + 0.59 * original_color[1] + 0.11 * original_color[2])

            # Calculate the new color
            r = int((1 - blend_factor) * original_color[0] + blend_factor * gray_value)
            g = int((1 - blend_factor) * original_color[1] + blend_factor * gray_value)
            b = int((1 - blend_factor) * original_color[2] + blend_factor * gray_value)

            # Update the pixel value
            grayscale_array[y, x] = [r, g, b, original_color[3]]

    # Conversion of the numpy array to an image
    grayscale_img = Image.fromarray(grayscale_array, "RGBA")

    # Save the grayscale image
    output_file = utils.create_output_file(outputs_path)
    utils.save_image_as_jpeg(output_file, grayscale_img.convert("RGB"))

    print("Grayed image saved to:", output_file)
    print("Execution time:", time.time() - start_time, "seconds")
